main_test.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"math/rand"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/google/uuid"
+)
+
+func TestGenerate(t *testing.T) {
+	uuid.SetRand(rand.New(rand.NewSource(1)))
+	in := configInput{
+		xdsServerUri:     "example.com:443",
+		gcpProjectNumber: 123456789012345,
+		vpcNetworkName:   "thedefault",
+		ip:               "10.9.8.7",
+		zone:             "uscentral-5",
+	}
+	expected := `{
+  "xds_servers": [
+    {
+      "server_uri": "example.com:443",
+      "channel_creds": [
+        {
+          "type": "google_default"
+        }
+      ]
+    }
+  ],
+  "node": {
+    "id": "52fdfc07-2182-454f-963f-5f0f9a621d72~10.9.8.7",
+    "cluster": "cluster",
+    "metadata": {
+      "TRAFFICDIRECTOR_GCP_PROJECT_NUMBER": "123456789012345",
+      "TRAFFICDIRECTOR_NETWORK_NAME": "thedefault"
+    },
+    "locality": {
+      "zone": "uscentral-5"
+    }
+  }
+}`
+	out, err := generate(in)
+	if err != nil {
+		t.Fatalf("expected no error, got: %v", err)
+	}
+	if expected != string(out) {
+		t.Fatalf("\nexpected:\n%v\n----------------\ngot:\n%v",
```

Comment:
This is going to be hard to make sense of if it fails.

One option would be to do a text `diff`, but it seems there is no canonical package for that.  Another option is to JSON parse and use https://godoc.org/github.com/google/go-cmp/cmp#Diff.  Yet another option would be to make sure the test never fails.

Reply:
I know it is hard to read. Initially I didn't set the random seed for UUID. It took me quite a bit of time to track down the problem to the UUID changing (it was the first time it came close to looking correct, so I assumed indentation was wrong or some such, because I had copied-and-pasted from a previous test run's output). I poked around and didn't see many alternatives.

I like the idea of cmp.Diff, although if it finds them to be equal and but the raw string is different I think I'll still output this hard-to-read version. I'll fix that up tomorrow.